### ISSUE TITLE
How to handle non-blocking async operations with coro in ambiorix?

### ISSUE THREAD

**Issue by spirae-nevin** [ambiorix-web/ambiorix#74]
**Problem** : I'm trying to make a long running task non blocking using `async` from `coro` in `ambiorix` but still getting blocked / delayed even though the function is asynchronously declared.

**Goal** : To let the server handle other requests in the meantime.

**Reproducible Example :**  
_app.R_
```
box::use(
  . / router[router],
)

box::use(
  ambiorix[Ambiorix]
)


app <- Ambiorix$new()
app$use(router)


app$start(port = 2050, host = "127.0.0.1")
```

_router.R_

```
box::use(
  ambiorix[Router]
)

box::use(
  . / controller[servePingInfo, serveLongRunningTask]
)

#' @export
router <- Router$new("/")

router$get("/ping", servePingInfo)
router$get("/long-task", serveLongRunningTask)

```

_controller.R_
```
box::use(
  ambiorix[Router],
  coro[async]
)

#' @export
servePingInfo <- async(function(req, res) {
  status <- "sucess"
  res$json(status)
})

#' @export
serveLongRunningTask <- async(function(req, res) {
  Sys.sleep(10)
  status <- "YOur wait is over"
  res$json(status)
})
```

**Current Behavior** : When hitting the /long-task endpoint, the server blocks for 10 seconds due to `Sys.sleep(10)` before responding. During this time, no other requests can be handled.

**Note** : It would be helpful if asynchronous approach can be done using `coro` libraries as it looks more similar to JavaScript.
Main thing to notice is that I'm not using `await()`. I cannot use await here like this router$get("/ping", await(servePingInfo)). This throws an error

**Comment by jrosell** [comment]
I have though a little for async in ambiorix too. I think it's worth exploring similar approach as ExtendedTask + furure/mirai.
Correct me if i'm wrong @JohnCoene, I think the current approach is using belgic.

**Comment by JohnCoene** [comment]
The way we have async is just using `future`, see [here](https://ambiorix.dev/docs/ambiorix/async), we can consider support for coro, i just need to understand the library a bit more to know what needs adapting in Ambiorix.

Belgic is a vertical load balancer for ambiorix, similar to shiny server.

**Comment by kennedymwavu** [comment]
hi @spirae-nevin,
can you provide a minimal reprex showing how to use {coro} for non-blocking async operations? (without ambiorix, just in a plain R session). would be happy to help!

**Comment by nev-awaken** [comment]
The `coro`  library has examples : https://coro.r-lib.org/reference/async.html

**Comment by kennedymwavu** [comment]
hi @spirae-nevin,
sorry for the late response (you never know when a promise will get resolved, huh?)
anyway, i've come back to this issue at least 10 times within the course of the past 1 year.
i didn't understand what {coro} was doing or why it was needed.
i took a look at it's docs again today, and funnily enough it just... clicked?! so, here we go:

## tl;dr

{coro} **doesn't** make blocking code non-blocking just because you wrapped it in `async()`.
it only becomes non-blocking when the async function **yields control back** to the event
loop by hitting an `await()` on something that is actually asynchronous (a promise, future, coro async op).

in other words, {coro} itself **does not provide concurrency**. it only provides **syntax and
control-flow** (`async()`, `await()`).

in your example, `serveLongRunningTask()` already violates the statement above: it has no `await()` on something that's async.

```r
#' @export
serveLongRunningTask <- async(function(req, res) {
  Sys.sleep(10)
  status <- "YOur wait is over"
  res$json(status)
})
```

this, obviously, results in a blocking operation for your whole R process.

## example

consider this function which returns a promise that resolves after one second:

```r
#' A Real Async Function
#'
#' @param x Integer /// Required.
#'        Hard-coded value that the promise will resolve to.
#'
#' @param delay Integer /// Optional.
#'              Number of seconds to delay promise resolution
#'              with. Default is 1.
#'
#' @return A promise that resolves after `delay` seconds.
#'
#' @export
delayed_value <- function(x, delay = 1L) {
  promises::promise(
    action = function(resolve, reject) {
      later::later(
        func = function() {
          message("resolved: ", x)
          resolve(x)
        },
        delay = delay
      )
    }
  )
}
```

suppose you want this workflow asynchronously:
1. get 10
2. add 5
3. multiply by 2
4. return the final result

with plain promises this is what you'd do:

```r
y <- delayed_value(x = 10)$then(
  function(x) {
    delayed_value(x + 5)
  }
)$then(
  function(x) {
    delayed_value(x * 2)
  }
)$then(
  function(x) {
    message("final result: ", x)
    x
  }
)
```

as you can see,  the logic is spread across callbacks.
but with {coro}, this is what you'd do:

```r
workflow <- coro::async(
  fn = function() {
    x <- coro::await(delayed_value(10))
    x <- coro::await(delayed_value(x + 5))
    x <- coro::await(delayed_value(x * 2))
    message("final result: ", x)
    x
  }
)

z <- workflow()
```

now it reads like normal sequential code.

## example ambiorix app

since we've established {coro} is only for syntax and control-flow, it means in real life you'll mostly use
it together with {future} or {mirai}.

but to start off with the usual `Sys.sleep()` example, what we need is an async sleep.

### async sleep

using plain promises a handler for async sleep would look like this:

```r
slow_get <- function(req, res) {
  promises::promise(
    action = function(resolve, reject) {
      later::later(
        func = function() {
          resolve(
            res$send("finally, we're done!")
          )
        },
        delay = 10L
      )
    }
  )
}
```

and here's the full reprex:

```r
library(promises)
library(ambiorix)

now <- function() {
  format(
    x = Sys.time(),
    format = "%F %T",
    usetz = TRUE
  )
}

home_get <- function(req, res) {
  res$send("<h1>hello, there!</h1>")
}

fast_get <- function(req, res) {
  msg <- paste("The time now is", now())

  res$send(msg)
}

slow_get <- function(req, res) {
  promises::promise(
    action = function(resolve, reject) {
      later::later(
        func = function() {
          msg <- paste("finally, we're done! it's now", now())
          resolve(
            res$send(msg)
          )
        },
        delay = 10L
      )
    }
  )
}

app <- Ambiorix$new()

app$get("/", home_get)
app$get("/fast", fast_get)
app$get("/slow", slow_get)

app$start(port = 5000L)
```

on the other hand, using {coro} the handler would be:

```r
slow_get <- async(function(req, res) {
  await(async_sleep(10))

  msg <- paste("finally, we're done! it's now", now())

  res$send(msg)
})
```

that's because {coro} already has `coro::async_sleep()`.
here's the full reprex:

```r
library(coro)
library(ambiorix)

now <- function() {
  format(
    x = Sys.time(),
    format = "%F %T",
    usetz = TRUE
  )
}

home_get <- function(req, res) {
  res$send("<h1>hello, there!</h1>")
}

fast_get <- function(req, res) {
  msg <- paste("The time now is", now())

  res$send(msg)
}

slow_get <- async(function(req, res) {
  await(async_sleep(10))

  msg <- paste("finally, we're done! it's now", now())

  res$send(msg)
})

app <- Ambiorix$new()

app$get("/", home_get)
app$get("/fast", fast_get)
app$get("/slow", slow_get)

app$start(port = 5000L)
```

### `Sys.sleep()`

R is single-threaded. `Sys.sleep()` is a blocking operation
on that thread, meaning no other request can be processed during that
time.

therefore, we must make `Sys.sleep()` and similar blocking work to run
on a separate R session or process, so that the main R session can continue
serving new requests.

now it's clear why we need to pair `{coro}`  with either `{future}` or `{mirai}`.

using `{future}` a handler would be something like:

```r
slow_get <- async(function(req, res) {
  x <- await(
    x = future::future(
      expr = {
        Sys.sleep(3L)
        10
      }
    )
  )

  y <- await(
    x = future::future(
      expr = {
        Sys.sleep(2L)
        x + 5
      }
    )
  )

  z <- await(
    x = future::future(
      expr = {
        Sys.sleep(1L)
        y * 2
      }
    )
  )

  msg <- sprintf(
    "final value: %.f, it's now: %s",
    z,
    now()
  )

  res$send(msg)
})
```

here's the reprex:

```r
library(coro)
library(future)
library(ambiorix)

future::plan(strategy = future::multisession)

now <- function() {
  format(
    x = Sys.time(),
    format = "%F %T",
    usetz = TRUE
  )
}

home_get <- function(req, res) {
  res$send("<h1>hello, there!</h1>")
}

fast_get <- function(req, res) {
  msg <- paste("The time now is", now())

  res$send(msg)
}

slow_get <- async(function(req, res) {
  x <- await(
    x = future::future(
      expr = {
        Sys.sleep(3L)
        10
      }
    )
  )

  y <- await(
    x = future::future(
      expr = {
        Sys.sleep(2L)
        x + 5
      }
    )
  )

  z <- await(
    x = future::future(
      expr = {
        Sys.sleep(1L)
        y * 2
      }
    )
  )

  msg <- sprintf(
    "final value: %.f, it's now: %s",
    z,
    now()
  )

  res$send(msg)
})

app <- Ambiorix$new()

app$get("/", home_get)
app$get("/fast", fast_get)
app$get("/slow", slow_get)

app$start(port = 5000L)
```

## conclusion

tbf, i was the primary target of this response (very selfish, ik). enjoyed every bit
of the research.

what i did there was blast you with a copious amount of code hoping you're going
to read and manually type it in your code editor (it's how i do it).
 
`{coro}` is only used for control flow to make async code read like synchronous code.
to actually perform async ops, you need to use it with `{future}` or `{mirai}`.
